cpp/src/iterator/iterator.cuh

```diff
@@ -0,0 +1,389 @@
+ *     using T_index = gdf_index_type;
+ *     make_iterator<has_nulls, T, T, T_transformer, T_index*>(...);
+ * -------------------------------------------------------------------------**/
+
+#ifndef CUDF_ITERATOR_CUH
+#define CUDF_ITERATOR_CUH
+
+#include <cudf.h>
+#include <bitmask/bit_mask.cuh>         // need for bit_mask::bit_mask_t
+#include <utilities/cudf_utils.h>       // need for CUDA_HOST_DEVICE_CALLABLE
+#include <utilities/error_utils.hpp>
+#include <utilities/type_dispatcher.hpp>
+
+#include <thrust/iterator/iterator_adaptor.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/pair.h>
+
+namespace cudf
+{
+namespace detail
+{
+
+/** -------------------------------------------------------------------------*
+ * @brief intermediate struct to calculate mean and variance
+ * This is an example case to output a struct from column input.
+ *
+ * this will be used to calculate and hold `sum of values`, 'sum of squares',
+ * 'sum of valid count'.
+ * Those will be used to compute `mean` (= sum / count)
+ * and `variance` (= sum of squares / count - mean^2).
+ *
+  @tparam  T  a element data type of value and value_squared.
+ * -------------------------------------------------------------------------**/
+template<typename T>
+struct meanvar
+{
+    T value;                /// the value
+    T value_squared;        /// the value of squared
+    gdf_index_type count;   /// the count
+
+    CUDA_HOST_DEVICE_CALLABLE
+    meanvar(T _value=0, T _value_squared=0, gdf_index_type _count=0)
+    : value(_value), value_squared(_value_squared), count(_count)
+    {};
+
+    using this_t = cudf::detail::meanvar<T>;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    this_t operator+(this_t const &rhs) const
+    {
+        return this_t(
+            (this->value + rhs.value),
+            (this->value_squared + rhs.value_squared),
+            (this->count + rhs.count)
+        );
+    };
+
+    CUDA_HOST_DEVICE_CALLABLE
+    bool operator==(this_t const &rhs) const
+    {
+        return (
+            (this->value == rhs.value) &&
+            (this->value_squared == rhs.value_squared) &&
+            (this->count == rhs.count)
+        );
+    };
+};
+
+// ---------------------------------------------------------------------------
+// transformers: transforms thrust::pair<T_element, bool> into output form.
+
+/** -------------------------------------------------------------------------*
+ * @brief transformer to output a scalar value
+ * A transformer for `column_input_iterator`
+ * It transforms `thrust::pair<T_element, bool>` into `T_output` form.
+ *
+ * This struct transforms the output value as `static_cast<T_output>(value)`.
+ *
+ * @tparam  T_element a scalar data type of input
+ * @tparam  T_output  a scalar data type of output
+ * -------------------------------------------------------------------------**/
+template<typename T_element, typename T_output=T_element>
+struct transformer
```

Comment:
I'd call this the `scalar_cast_transformer` as it just casts the scalar from one type to another.